### PR TITLE
Infer language for code snippets in descriptions.

### LINKED ISSUE #18: Add auto syntax-highlighting for code snippets
(I hope this is the correct place for this request)

When reading the autogenerated docs, the snippets are a bit hard to read/extract from the other documentation. Syntax highlighting could help with this problem. The visual difference of the highlighting makes distinguishing code snippets from pure docs easier. It would be nice if the bot could detect these snippets and automatically add &#96;&#96;&#96;HTML in front of it and closes it with &#96;&#96;&#96;.

For example in the `paper-input` docs replace

```
<paper-input-container>
  <input is="iron-input" maxlength="20">
  <paper-input-char-counter></paper-input-char-counter>
</paper-input-container>
```

with

``` HTML
<paper-input-container>
  <input is="iron-input" maxlength="20">
  <paper-input-char-counter></paper-input-char-counter>
</paper-input-container>
```

In markdown code:
&#96;&#96;&#96;HTML
&lt;paper-input-container>
&nbsp;&nbsp;&lt;input is="iron-input" maxlength="20">
&nbsp;&nbsp;&lt;paper-input-char-counter>&lt;/paper-input-char-counter>
&lt;/paper-input-container>
&#96;&#96;&#96;

### PR DESCRIPTION
**Pull request by rictic**
Fixes #18

### CODE REVIEW

File: markdown-lang-autodetect.ts [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    // Here we automatically detect the language for the given code snippet
+    // and inject it into the output markdown.
+    if (!lang) {
+      lang = highlight.highlightAuto(
+          code, ['html', 'css', 'javascript']).language;
+    }
+    if (escaped) {
+      code = entities.decode(code);
+    }
+    return `\`\`\`${lang}\n${code}\n\`\`\`\n\n`;
```
Comment (dfreedm) [inline markdown-lang-autodetect.ts:46]: This is the saddest line :cry:

File: markdown-lang-autodetect.ts [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+  paragraph(text:string) {
+    return `${text}\n\n`;
+  }
+
+  table(header:string, body:string) {
+    return `${header}\n${body}\n`;
+  }
+
+  tablerow(content:string) {
+    const isHeader = /😿TH😿/.test(content);
```
Comment (dfreedm) [inline markdown-lang-autodetect.ts:93]: These aren't meant to be crying cats right?

File: markdown-lang-autodetect.ts (hunk 1 above)
Comment (rictic) [inline markdown-lang-autodetect.ts:46]: \W\h\at\ \d\o\ \y\o\u\ \m\e\a\n\?

File: markdown-lang-autodetect.ts (hunk 2 above)
Comment (rictic) [inline markdown-lang-autodetect.ts:93]: That's a good question, and a fair one. I'm happy to report that they are, in fact, meant to be crying cats. I figured that they make a good separator character when you don't want to bother with escaping.

They're inserted and removed by the script to tell whether a row is a table header or just a normal row. (the renderer API is designed to represent html more than it is to represent markdown sadly, so it gives you header info at the cell level but we need it at the row level. Likewise not giving us nesting info for lists)

Comment (rictic) [inline markdown-lang-autodetect.ts:93]: I could do it the proper way and escape the table header symbol, I was just lazy last night.

File: custom_typings/highlight.js.d.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
+declare module 'highlight.js' {
+  interface HighlightAutoResult {
+    /** detected language */
+    language:string;
```
Comment (dfreedm) [inline custom_typings/highlight.js.d.ts]: nit space after `:`

File: custom_typings/html-entities.d.ts [hunk 4]
```diff
@@ -0,0 +1,5 @@
+declare module 'html-entities' {
+  export var AllHtmlEntities: {
+    decode(text:string):string;
```
Comment (dfreedm) [inline custom_typings/html-entities.d.ts]: nit space after `:`

File: markdown-lang-autodetect.ts (hunk 2 above)
Comment (dfreedm) [inline markdown-lang-autodetect.ts:93]: Nah, that seems fine for now.

**Comment by dfreedm** [comment]
LGTM with nits

File: markdown-lang-autodetect.ts (hunk 2 above)
Comment (TimvdLippe) [inline markdown-lang-autodetect.ts:93]: :joy_cat: :joy_cat: :joy_cat: Had a good laugh reading this.

P.S.: Looking forward to this change, thanks for doing it!

File: custom_typings/html-entities.d.ts (hunk 4 above)
Comment (rictic) [inline custom_typings/html-entities.d.ts]: Done

File: custom_typings/highlight.js.d.ts (hunk 3 above)
Comment (rictic) [inline custom_typings/highlight.js.d.ts]: Done

**Pull request merged**